packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp

```diff
@@ -654,54 +768,104 @@ class ThyraProductME_Constraint_SimOpt : public Constraint_SimOpt<Real> {
+      value_ptr_ = c.clone();
+    value_ptr_->set(c);
+
+    computeValue = solveConstraint = false;
   }
 
 
   /** \brief Update constraint functions with respect to Sim variable.
                 x is the optimization variable,
-                flag = true if optimization variable is changed,
+                flag = ??,
                 iter is the outer algorithm iterations count.
    */
-  void update_1( const Vector<Real> &u, bool flag = true, int iter = -1 ) {
-    if (flag == true) {
-      updateValue = true;
-      updateJacobian1 = true;
+  void update_1( const Vector<Real> &u, bool /*flag*/ = true, int iter = -1 ) {
+    if(u_hasChanged(u)) {
+      if(verbosityLevel >= Teuchos::VERB_HIGH)
+        *out << "ROL::ThyraProductME_Constraint_SimOpt::update_1, The State Changed" << std::endl;
+      computeValue = computeJacobian1 = true;
+
+      if (Teuchos::is_null(rol_u_ptr))
+        rol_u_ptr = u.clone();
+      rol_u_ptr->set(u);
     }
+
     if(params != Teuchos::null)
       params->set<int>("Optimizer Iteration Number", iter);
   }
 
   /** \brief Update constraint functions with respect to Opt variable.
                 x is the optimization variable,
-                flag = true if optimization variable is changed,
+                flag = ??,
                 iter is the outer algorithm iterations count.
    */
-  void update_2( const Vector<Real> &z, bool flag = true, int iter = -1 ) {
-    if (flag == true) {
-      updateValue = true;
-      updateJacobian1 = true;
+  void update_2( const Vector<Real> &z, bool /*flag*/ = true, int iter = -1 ) {
+    if(z_hasChanged(z)) {
+      if(verbosityLevel >= Teuchos::VERB_HIGH)
+        *out << "ROL::ThyraProductME_Constraint_SimOpt::update_2, The Parameter Changed" << std::endl;
+      computeValue = computeJacobian1 = solveConstraint = true;
+
+      if (Teuchos::is_null(rol_z_ptr))
+        rol_z_ptr = z.clone();
+      rol_z_ptr->set(z);
     }
+
     if(Teuchos::nonnull(params)) {
+      auto& z_stored_ptr = params->get<Teuchos::RCP<Vector<Real> > >("Optimization Variable");
+      if(Teuchos::is_null(z_stored_ptr) || z_hasChanged(*z_stored_ptr)) {
+        if(verbosityLevel >= Teuchos::VERB_HIGH)
+          *out << "ROL::ThyraProductME_Constraint_SimOpt::update_2, Signaling That Parameter Changed" << std::endl;
+        params->set<bool>("Optimization Variables Changed", true);
+        if(Teuchos::is_null(z_stored_ptr))
+          z_stored_ptr = z.clone();
+        z_stored_ptr->set(z);
+      }
+
       params->set<int>("Optimizer Iteration Number", iter);
-      if(flag == true)
-        params->set<bool>("Optimization Variables Changed",true);
     }
   }
 
+  bool z_hasChanged(const Vector<Real> &rol_z) const {
+    bool changed = true;
+    if (Teuchos::nonnull(rol_z_ptr)) {
+      auto diff = rol_z.clone();
+      diff->set(*rol_z_ptr);
+      diff->axpy( -1.0, rol_z );
+      Real norm = diff->norm();
+      changed = (norm == 0) ? false : true;
```

Comment:
Should this check have a tolerance due to roundoff?

Reply:
No.. I really want the two field to be exactly the same, in which case `diff` is a vector of zeros and  norm will be zero with no roundoff errors.